homeassistant/components/tplink/entity.py

```diff
@@ -62,24 +108,272 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+        # platform is added it should provide an implementation here to
+        # return legacy_device_id(self._device)
+        raise NotImplementedError
+
+    @abstractmethod
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Platforms implement this to update the entity internals."""
+        raise NotImplementedError
+
+    @callback
+    def _async_call_update_attrs(self) -> None:
+        """Call update_attrs and make entity unavailable on error.
+
+        update_attrs can sometimes fail if a device firmware update breaks the
+        downstream library.
+        """
+        try:
+            self._async_update_attrs()
+        except Exception as ex:  # noqa: BLE001
+            if self._attr_available:
+                _LOGGER.warning(
+                    "Unable to read data for %s %s: %s",
+                    self._device,
+                    self.entity_id,
+                    ex,
+                )
+            self._attr_available = False
+        else:
+            self._attr_available = True
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._async_call_update_attrs()
+        super()._handle_coordinator_update()
+
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        return self.coordinator.last_update_success and self._attr_available
+
+
+class CoordinatedTPLinkFeatureEntity(CoordinatedTPLinkEntity, ABC):
+    """Common base class for all coordinated tplink feature entities."""
+
+    entity_description: TPLinkFeatureEntityDescription
+    _feature: Feature
+
+    def __init__(
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        *,
+        feature: Feature,
+        description: TPLinkFeatureEntityDescription,
+        parent: Device | None = None,
+    ) -> None:
+        """Initialize the entity."""
+        self.entity_description = description
+        super().__init__(device, coordinator, parent=parent, feature=feature)
+
+    def _get_unique_id(self) -> str:
+        """Return unique ID for the entity."""
+        key = self.entity_description.key
+        if key == PRIMARY_STATE_ID:
+            return legacy_device_id(self._device)
+
+        key = LEGACY_KEY_MAPPING.get(key, key)
+        return f"{legacy_device_id(self._device)}_{key}"
+
+
+def _entities_for_device[_E: CoordinatedTPLinkEntity](
+    device: Device,
+    coordinator: TPLinkDataUpdateCoordinator,
+    *,
+    feature_type: Feature.Type,
+    entity_class: type[_E],
+    parent: Device | None = None,
+) -> list[_E]:
+    """Return a list of entities to add.
+
+    This filters out unwanted features to avoid creating unnecessary entities
+    for device features that are implemented by specialized platforms like light.
+    """
+    entities: list[_E] = []
+
+    entities.extend(
```

Comment:
```suggestion
    entities: list[_E] = [
```

Reply:
Could be a direct list comp

Reply:
Done in https://github.com/home-assistant/core/pull/120064